### DISCUSSION TITLE
Reliably get the Gdk Monitor when shown

### DISCUSSION THREAD

**Discussion by ErikReider** [wmww/gtk4-layer-shell#22] [Q&A]
Hi! Thanks for your amazing work!

I've been encountering a little snag when porting from the GTK3 version to this one. In the previous release, I was able to get the current Monitor that the surface was attached to by listening to the "::has-toplevel-focus" signal after the "::map" signal which worked reliably (without running `set_monitor`):

```vala
// GTK3
this.map.connect (() => {
    ulong id = 0;
    id = notify["has-toplevel-focus"].connect (() => {
        disconnect (id);
        unowned Gdk.Monitor monitor = null;
        unowned Gdk.Window ? win = get_window ();
        if (win != null) {
            monitor = get_display ().get_monitor_at_window (win);
        }
    });
});
```

Now when the "::has-toplevel-focus" has been removed in GTK4, is there any of achieving this/a better way of doing this in GTK4?

Thanks in advance :)

**Comment by wmww** [comment]
Hmm, I'm not sure about that, but I suspect there's a way to make it work. Does `.get_monitor_at_window(win);` not work in the window's `map` callback? Perhaps you could simply call it after a short delay?

**Comment by wmww** [comment]
Oh, maybe this is what you want? https://docs.gtk.org/gdk4/signal.Surface.enter-monitor.html

**Reply by ErikReider** [reply]
Using this in the map callback works wonderfully! Thanks for your time :)

**Reply by ErikReider** [reply]
Seems like that signal is only called once. This is the Vala code used:

```vala
this.map.connect_after (() => {
    // Wait until the layer has attached
    unowned Gdk.Surface surface = get_surface ();
    if (!(surface is Gdk.Surface)) return;
    var mon = surface.get_display ().get_monitor_at_surface (surface);
    print ("Map monitor: %s\n", mon.connector);
    ulong id = 0;
    id = surface.enter_monitor.connect ((surface, monitor) => {
        print ("Enter: %s\n", monitor.connector);
        surface.disconnect (id);
    });
});
```

Here's what happens when I run the code above (hopefully this is clear enough):

The first time displaying the surface:
- Map check: null monitor
- Enter_monitor check: Correct monitor

The second time displaying the surface:
- Map check: Correct monitor
- Enter_monitor check: Isn't ran

Switching the focus to my secondary monitor:
- Map check: Previous focused monitor
- Enter_monitor check: Correct monitor

The second time displaying the surface on the second monitor:
- Map check: Correct monitor
- Enter_monitor check: Isn't ran

Switching back the focus to the first checked monitor:
- Map check: always my secondary monitor
- Enter_monitor check: Isn't ran
